### PR TITLE
feature: DynSolCall

### PR DESCRIPTION
**Pull request by prestwich**
## Motivation

Continue fleshing out dyn-abi to reach feature parity with sol-types

## Solution

- Add `DynSolCall`
- Add `DynSolReturns`
- Add `impl Specifier<DynSolCall> for Function`

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Comment by prestwich** [comment]
anyone know what's up with miri cancelling?

File: crates/dyn-abi/src/eip712/mod.rs [hunk 1]
```diff
@@ -16,3 +16,39 @@ mod resolver;
 pub use resolver::{PropertyDef, Resolver, TypeDef};
 
 pub(crate) mod coerce;
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use alloy_primitives::B256;
+    use alloy_sol_types::SolStruct;
+    #[test]
```
Comment (DaniPopes) [inline crates/dyn-abi/src/eip712/mod.rs:25]:
```suggestion

    #[test]
```

File: crates/dyn-abi/src/ext/abi.rs [hunk 2]
```diff
@@ -118,29 +118,29 @@ impl JsonAbiExt for Error {
...
     #[inline]
     fn abi_encode_output(&self, values: &[DynSolValue]) -> Result<Vec<u8>> {
-        encode_typeck(&self.outputs, values)
+        self.resolve()?.abi_encode_output(values)
     }
 
     #[inline]
     fn abi_decode_output(&self, data: &[u8], validate: bool) -> Result<Vec<DynSolValue>> {
-        abi_decode(data, &self.outputs, validate)
+        self.resolve()?.abi_decode_output(data, validate)
```
Comment (DaniPopes) [inline crates/dyn-abi/src/ext/abi.rs:143]: Can you please keep these specialized functions, we don't want to calculate the signature and collect inputs and outputs on every call

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/dyn-abi/src/ext/abi.rs (hunk 2 above)
Comment (prestwich) [reply crates/dyn-abi/src/ext/abi.rs:143]: hmmm. I think that if users are keeping the json representation in memory and passing it around, that's generally an anti-pattern, right?  I see these traits as convenience functions for one-off things and to patch functionality holes, not as core API. `Dyn____` is our in-memory representation of ABI information, and greedily resolving a `JsonAbi` into a `DynAbi` is the longterm officially-recommended path

Comment (prestwich) [reply crates/dyn-abi/src/ext/abi.rs]: reverting the file to unblock this

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**